File: src/main/java/rars/venus/settings/editor/TreePanel.java
package rars.venus.settings.editor;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.DefaultTreeCellRenderer;
import javax.swing.tree.MutableTreeNode;
import java.awt.*;

public final class TreePanel extends JScrollPane {
    public final @NotNull TreeNode fontSettingsNode, generalSchemeSettingsNode, syntaxSettingsNode, otherSettingsNode, presetsNode;

    public final @NotNull JTree tree;

    public TreePanel() {
        super();
        this.fontSettingsNode = new TreeNode.Normal("Font");
        this.generalSchemeSettingsNode = new TreeNode.Normal("General");
        this.syntaxSettingsNode = new TreeNode.Normal("RISC-V Syntax");
        this.otherSettingsNode = new TreeNode.Normal("Other settings");
        this.presetsNode = new TreeNode.Normal("Presets");
        this.tree = buildTree(fontSettingsNode, presetsNode, generalSchemeSettingsNode, syntaxSettingsNode, otherSettingsNode);
        this.setViewportView(this.tree);

        final var renderer = (DefaultTreeCellRenderer) this.tree.getCellRenderer();
        renderer.setLeafIcon(null); // Remove icons for leaf nodes
        renderer.setClosedIcon(null);
        renderer.setOpenIcon(null);

        // Keep the navigation panel compact vertically
        this.setPreferredSize(new Dimension(200, 300));
    }

    private static @NotNull JTree buildTree(
        final @NotNull TreeNode fontSettingsNode,
        final @NotNull TreeNode presetsNode,
        final @NotNull TreeNode generalSchemeSettingsNode,
        final @NotNull TreeNode syntaxSettingsNode,
        final @NotNull TreeNode otherSettingsNode
    ) {
        final var newRoot = root(
            presetNode(fontSettingsNode),
            normalNode("Color scheme").children(
                presetNode(presetsNode),
                presetNode(generalSchemeSettingsNode),
                presetNode(syntaxSettingsNode)
            ),
            presetNode(otherSettingsNode)
        ).collect();
        final var tree = new JTree(newRoot);
        tree.setRootVisible(false); // Hide the root node
        tree.setShowsRootHandles(true); // Show expand/collapse icons
        return tree;
    }

    @Contract("_ -> new")
    private static @NotNull NodeBuilder normalNode(final @NotNull String name) {
        return new NodeBuilder(new TreeNode.Normal(name));
    }

    private static @NotNull NodeBuilder presetNode(final @NotNull TreeNode node) {
        return new NodeBuilder(node);
    }

    private static @NotNull NodeBuilder root(final @NotNull NodeBuilder... children) {
        return new NodeBuilder(null).children(children);
    }

    private static final class NodeBuilder {
        private final DefaultMutableTreeNode node;

        public NodeBuilder(final Object data) {
            this.node = new DefaultMutableTreeNode(data);
        }

        public NodeBuilder children(final NodeBuilder @NotNull ... children) {
            for (final var child : children) {
                this.node.add(child.collect());
            }
            return this;
        }

        public MutableTreeNode collect() {
            return node;
        }
    }
}

File: src/main/java/rars/venus/settings/editor/TreePanel.kt
package rars.venus.settings.editor

import rars.settings.TokenSettingKey
import java.awt.Dimension
import javax.swing.JScrollPane
import javax.swing.JTree
import javax.swing.tree.DefaultMutableTreeNode
import javax.swing.tree.DefaultTreeCellRenderer
import javax.swing.tree.TreeNode

class TreePanel : JScrollPane() {
    val fontSettingsNode: TreeNodeData = TreeNodeData("Font")
    val generalSchemeSettingsNode: TreeNodeData = TreeNodeData("General")
    val otherSettingsNode: TreeNodeData = TreeNodeData("Other settings")
    val presetsNode: TreeNodeData = TreeNodeData("Presets")

    val tree = buildTree(
        fontSettingsNode,
        presetsNode,
        generalSchemeSettingsNode,
        otherSettingsNode
    )

    init {
        setViewportView(tree)

        (tree.cellRenderer as DefaultTreeCellRenderer).apply {
            leafIcon = null
            closedIcon = null
            openIcon = null
        }

        preferredSize = Dimension(200, 400)
    }
}

private fun buildTree(
    fontSettingsNode: TreeNodeData,
    presetsNode: TreeNodeData,
    generalSchemeSettingsNode: TreeNodeData,
    otherSettingsNode: TreeNodeData
): JTree {
    val newRoot = buildRoot {
        presetNode(fontSettingsNode)
        normalNode("Color scheme") {
            presetNode(presetsNode)
            presetNode(generalSchemeSettingsNode)
            normalNode("RISC-V Syntax") {
                TokenSettingKey.entries.forEach { key ->
                    syntaxNode(key, key.description)
                }
            }
        }
        presetNode(otherSettingsNode)
    }

    return JTree(newRoot).apply {
        isRootVisible = false
        showsRootHandles = true
    }
}

private fun buildRoot(builder: DefaultMutableTreeNode.() -> Unit): TreeNode =
    DefaultMutableTreeNode(null).apply(builder)

private fun DefaultMutableTreeNode.normalNode(
    name: String,
    builderFunc: DefaultMutableTreeNode.() -> Unit
) {
    DefaultMutableTreeNode(TreeNodeData(name)).apply(builderFunc)
        .also { add(it) }
}

private fun DefaultMutableTreeNode.presetNode(node: TreeNodeData) {
    DefaultMutableTreeNode(node).also { add(it) }
}

private fun DefaultMutableTreeNode.syntaxNode(
    type: TokenSettingKey,
    name: String
) {
    DefaultMutableTreeNode(TreeNodeData.Syntax(type, name)).also { add(it) }
}

